src/core/qgsmaprenderertask.cpp

```diff
@@ -149,31 +170,45 @@ bool QgsMapRendererTask::run()
   if ( !mFileName.isEmpty() )
   {
     destPainter->end();
-    bool success = img.save( mFileName, mFileFormat.toLocal8Bit().data() );
-    if ( !success )
+
+    if ( mForceRaster && mFileFormat == QStringLiteral( "PDF" ) )
     {
-      mError = ImageSaveFail;
-      return false;
+      QPainter pp;
+      pp.begin( printer );
+      QRectF rect( 0, 0, img.width(), img.height() );
+      pp.drawImage( rect, img, rect );
+      pp.end();
     }
-
-    if ( mSaveWorldFile )
+    else if ( mFileFormat != QStringLiteral( "PDF" ) )
     {
-      QFileInfo info  = QFileInfo( mFileName );
-
-      // build the world file name
-      QString outputSuffix = info.suffix();
-      QString worldFileName = info.absolutePath() + '/' + info.baseName() + '.'
-                              + outputSuffix.at( 0 ) + outputSuffix.at( info.suffix().size() - 1 ) + 'w';
-      QFile worldFile( worldFileName );
+      bool success = img.save( mFileName, mFileFormat.toLocal8Bit().data() );
+      if ( !success )
+      {
+        mError = ImageSaveFail;
+        return false;
```

Comment:
This method will leak `printer` if you early exit here.
Just make it a `std::unique_ptr`